### PR TITLE
Measure writing guide: add section about I/O file + modernize 

### PR DESCRIPTION
**Pull request by jmarrec**
Measure writing guide: add section about I/O file + modernize 

Fixes https://github.com/NREL/OpenStudio/issues/4046

Fixes https://github.com/NREL/OpenStudio/issues/3609

### CODE REVIEW

File: docs/reference/measure_writing_guide.md [hunk 1]
```diff
@@ -55,7 +55,13 @@ end
...
 ```
 
-Please note the original measure instantiantiation syntax, ```OpenStudio::Ruleset::ModelUserScript```, is deprecated.
+Please note the original namespace in versions of OpenStudio < 2.X `OpenStudio::Ruleset` is deprecated and replaced with `OpenStudio::Measure`.
+Measure classes prior to 2.X are also deprecated:
+
+| < 2.X (Deprecated)                       | >= 2.X                            |
+|------------------------------------------|-----------------------------------|
+| OpenStudio::Ruleset::ModelUserScript     | OpenStudio::Measure::ModelMeasure |
+| OpenStudio::Ruleset::WorkspaceUserScript | OpenStudio::Measure::ModelMeasure |
```
Comment (jmarrec) [inline docs/reference/measure_writing_guide.md:64]: I did some changed re OpenStudio::Ruleset > OpenStudio::Measure

File: docs/reference/measure_writing_guide.md [hunk 2]
```diff
@@ -120,8 +126,8 @@ We define a user argument vector and create a single argument for insulation thi
 
 ```ruby
 def arguments(model)
-    args = OpenStudio::Ruleset::OSArgumentVector.new
-    insl_thckn = OpenStudio::Ruleset::makeDoubleArgument('insl_thckn',true)
+    args = OpenStudio::Measure::OSArgumentVector.new
+    insl_thckn = OpenStudio::Measure::makeDoubleArgument('insl_thckn',true)
```
Comment (jmarrec) [inline docs/reference/measure_writing_guide.md:130]: One of many

File: docs/reference/measure_writing_guide.md [hunk 3]
```diff
@@ -1466,6 +1473,12 @@ Duplicate requests are ignored; in the case of unique objects, the requests are
   end
 ```
 
+### Reporting Measure arguments Method
+
+```ruby
+def arguments(model = nil)
+```
+
```
Comment (jmarrec) [inline docs/reference/measure_writing_guide.md:1481]: Added

File: docs/reference/measure_writing_guide.md [hunk 4]
```diff
@@ -1553,6 +1566,38 @@ An attribute named 'rotation' will automatically be added to the measure's outpu
...
+    File.open('./qa_qc.csv', 'w') do |file|
+      # Write file
+    end
+  end
+end
+
+FooBar.new.registerWithApplication
+```
+
+
```
Comment (jmarrec) [inline docs/reference/measure_writing_guide.md:1600]: Document the classic case: output a report file from a ReportingMeasure.

File: docs/reference/measure_writing_guide.md [hunk 5]
```diff
@@ -1694,6 +1741,134 @@ Below is an example run method that manipulates and reports values related to a
...
+    end
+
+    column = 1
+    rowsToSkip = 1
+    scheduleFile = OpenStudio::Model::ScheduleFile.new(externalFile_.get(), column, rowsToSkip)
+    scheduleFile.setName("ExampleScheduleFile")
+
+    return true
+end
+```
```
Comment (jmarrec) [inline docs/reference/measure_writing_guide.md:1798]: This section is fair game and general enough. I definitely think it should be added.

File: docs/reference/measure_writing_guide.md [hunk 6]
```diff
@@ -1694,6 +1741,134 @@ Below is an example run method that manipulates and reports values related to a
...
+
+#### How to output any other file
+
+To output any file you may need that isn't an `ExternalFile`, you should rely on two things, in order of preferences:
+* `runner.workflow.filePaths[0]`: this will typically resolve to `./generated_files`
+* `runner.workflow.absoluteRootDir`: this will resolve to '.'
+
+`.` is the location defined as the `root` key inside the `workflow.osw`, or if not specified the location of the `workflow.osw` itself.
```
Comment (jmarrec) [inline docs/reference/measure_writing_guide.md:1806]: I find that relevant, but it's possible other people might not?

File: docs/reference/measure_writing_guide.md [hunk 7]
```diff
@@ -1694,6 +1741,134 @@ Below is an example run method that manipulates and reports values related to a
...
+  def run(model, runner, user_arguments)
+
+    # Locate the resource file (this resolves to something like './measures/resources/schedulefile.csv')
+    csv_in_path = File.join(File.dirname(__FILE__), 'schedulefile.csv')
+
+    # Canonical way: write to the ./generated_files directory
+    out_file = File.join(runner.workflow.filePaths[0].to_s, 'myfile.csv')
+    File.open(out_file, 'w') do |f|
+      f << "Hello"
+    end
```
Comment (jmarrec) [inline docs/reference/measure_writing_guide.md:1822]: I think that part is definitely fine.

File: docs/reference/measure_writing_guide.md [hunk 8]
```diff
@@ -1694,6 +1741,134 @@ Below is an example run method that manipulates and reports values related to a
...
+      html_out_path = File.absolute_path(
+        File.join(rootDir, 'reports', html_out_path))
+    end
+    outDir = File.dirname(html_out_path)
+    if !File.exists?(outDir)
+      FileUtils.mkdir_p(outDir)
+    end
+    File.open(html_out_path, 'w') do |f|
+      f << "<html><head><title>My Custom Report that works with Apply Now!</title></head><body><h1>Hello World!</h1></body></html>"
+    end
```
Comment (jmarrec) [inline docs/reference/measure_writing_guide.md:1846]: So perhaps this section is overkill or an invitation for bad ideas. I don't know. But people are definitely already doing stuff like this (cf discussion on https://github.com/NREL/OpenStudio/issues/4046) except they are not doing it right for lack of documentation.

File: docs/reference/measure_writing_guide.md [hunk 9]
```diff
@@ -1694,6 +1741,134 @@ Below is an example run method that manipulates and reports values related to a
...
+
+```ruby
+    if runner.workflow.currentStepIndex() > 0
+      previousStep = runner.workflow.workflowSteps[runner.workflow.currentStepIndex() - 1]
+      if previousStep.result
+        previousStepResult = previousStep.result.get
+        runner.registerWarning("previousStepResult.stepFiles=#{previousStepResult.stepFiles.map{|p| p.to_s}}")
+      end
+    end
+```
```
Comment (jmarrec) [inline docs/reference/measure_writing_guide.md:1869]: I don't have a good use case for this but perhaps @DavidGoldwasser might. I won't be offended if you don't lik that section

**Comment by tijcolem** [comment]
@jmarrec Changes look good.  Thanks for the updates.  I just assigned this to myself.  I'd like to build/convert the docs to html before merging to gh-pages branch.

**Comment by tijcolem** [comment]
Looks good served up locally using mkdocs.

**Pull request merged**